### PR TITLE
[`core` / `DDP`] Fix RM trainer + DDP + quantization + propagate `gradient_checkpointing_kwargs` in SFT & DPO

### PR DESCRIPTION
**Pull request by younesbelkada**
Needs : https://github.com/huggingface/peft/pull/1036 / https://github.com/huggingface/transformers/pull/27020 

Fixes: https://github.com/huggingface/trl/issues/891
Fixes: https://github.com/huggingface/trl/issues/835

To avoid issues with PEFT + DDP, we need to call the gradient checkpointing method with `use_reentrant=False` that you can pass over the argument `gradient_checkpointing_kwargs` directly in trainer with https://github.com/huggingface/transformers/pull/27020 .

For users that do not have the correct transformers version they need to update transformers with the correct version to get that feature, otherwise `gradient_checkpointing_kwargs` will get ignored.

cc @lvwerra

### CODE REVIEW

File: examples/scripts/reward_modeling.py [hunk 1]
```diff
@@ -103,7 +103,7 @@ class ScriptArguments:
 
 # Step 2: Load the dataset and pre-process it
 tokenizer = AutoTokenizer.from_pretrained(args.model_name)
-train_dataset = load_dataset(args.dataset_name, split="train")
+train_dataset = load_dataset(args.dataset_name, split="train[:50]")
```
Comment (vwxyzjn) [inline examples/scripts/reward_modeling.py:106]: Could this be not hardcoded?

**Review by vwxyzjn — COMMENTED** [review]
LG! One minor comment.

File: examples/scripts/reward_modeling.py [hunk 2]
```diff
@@ -98,7 +99,7 @@ class ScriptArguments:
...
```
Comment (younesbelkada) [inline examples/scripts/reward_modeling.py:102]:
```suggestion
train_dataset = load_dataset(args.dataset_name, split="train")
```
I used that for testing purpose, will remove it!
